### PR TITLE
TYP: fix typing error in hookspecs, refactor NebariStage

### PR DESCRIPTION
**Pull request by fangchenli**
## Reference Issues or PRs
xref #1800 ##1894

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

### CODE REVIEW

File: src/nebari/hookspecs.py [hunk 1]
```diff
@@ -1,54 +1,37 @@
...
-        self, stage_outputs: Dict[str, Dict[str, Any]], status: Dict[str, bool]
-    ):
-        yield
+    @abstractmethod
+    def render(self):
+        ...
 
 
 @hookspec
-def nebari_stage() -> List[NebariStage]:
```
Comment (fangchenli) [inline src/nebari/hookspecs.py:48]: Empty method body causes missing return error.

File: src/nebari/hookspecs.py [hunk 2]
```diff
@@ -1,54 +1,37 @@
...
-    def check(
-        self, stage_outputs: Dict[str, Dict[str, Any]], disable_prompt: bool = False
-    ) -> bool:
-        pass
-
-    @contextlib.contextmanager
-    def destroy(
-        self, stage_outputs: Dict[str, Dict[str, Any]], status: Dict[str, bool]
-    ):
-        yield
```
Comment (fangchenli) [inline src/nebari/hookspecs.py:44]: Those don't apply for BootstrapStage.

File: src/nebari/hookspecs.py [hunk 3]
```diff
@@ -1,54 +1,37 @@
...
-    input_schema: pydantic.BaseModel = None
-    output_schema: pydantic.BaseModel = None
-
-    def __init__(self, output_directory: pathlib.Path, config: schema.Main):
+    def __init__(self, output_directory: Path, config: schema.Main):
         self.output_directory = output_directory
         self.config = config
 
-    def render(self) -> Dict[str, str]:
-        return {}
```
Comment (fangchenli) [inline src/nebari/hookspecs.py:27]: The return type should be `Dict[Union[str, Path], str]`. Then we need to use the overload decorator in the derived class.

File: src/nebari/hookspecs.py (hunk 2 above)
Comment (iameskild) [reply src/nebari/hookspecs.py:44]: Removing these hookspec methods effectively removes these as pluggable methods for downstream extensions.

These methods are not used by the BootstrapStage but are used by NebariTerraformStage. 

Perhaps there is more work needed before we can tear these methods out. Either adding them as hookspecs in another class or something else.

File: src/nebari/hookspecs.py (hunk 3 above)
Comment (iameskild) [reply src/nebari/hookspecs.py:27]: Is this something you'll handle in this PR?

**Review by iameskild — CHANGES_REQUESTED** [review]
Thanks @fangchenli! I left a few comments that I think we should discuss a bit further.

File: src/nebari/hookspecs.py (hunk 3 above)
Comment (fangchenli) [reply src/nebari/hookspecs.py:27]: There is no elegant way to do this currently. I think it's better to left those abstract method without return time.

Comment (fangchenli) [reply src/nebari/hookspecs.py:27]: The return type should actually be `Dict[Path, str]`.

**Comment by iameskild** [comment]
@costrouc could you have a look at this PR?

**Comment by viniciusdc** [comment]
This has some big refactorings in the Nebari Stage which needs to be retested now that we changed a bit the behavior for the default schema dumps -- I will follow up on this, next week

**Comment by dcmcand** [comment]
@viniciusdc should we close this one or is there still value here?

**Comment by viniciusdc** [comment]
Hi @dcmcand, There is a lot of value here. This is good work that we will need to encompass sooner or later (related to that schema refactoring I was talking about recently). I was waiting on this due to the exposing default config values Adam was working on.

My opinion on this would be to close, and I will incorporate and cite those changes in a follow up PR

**Pull request closed without merging**